index.rst

```diff
@@ -1,25 +1,429 @@
+Reliability and performance can be increased by running multiple brokers in parallel, and by using more than one "partition" for each topic.
+
+Schema Evolution
+----------------
+
+DDS does not support schema evolution, at least as we are using it.
+When we update the schema for a topic it becomes an entirely new topic.
+Internally the topic has a different name; the topic names DDS sees include a hash that represents the schema
+though we hide this hash from users.
+
+Kafka schemas support basic schema evolution: we may add or delete fields and readers and writers can freely use older or newer schemas.
+Note that even with this type of schema evolution, we will have to reset part of the system if we change the type of a message field.
+
+Licensing
+---------
+
+Kafka is open-source, ADLink OpenSplice is not.
+With Kafka developers can run the full system, with all features, without licensing headaches.
+With ADLink OpenSplice we presently run two very different versions of the software, in order to reduce costs and avoid licensing headaches:
+
+* Developers use the “community” edition, which ADLink no longer offers, and which only supports single-process topic daemons.
+
+* Deployment uses the commercial edition, because it supports shared-memory topic daemons, which work better for large systems.
+  The commerical version is also required to run ADLink's diagnostic tools.
+  Note that the commercial edition has always been newer than the community edition.
+  In the past ADLink kept the community edition one major version behind the commercial edition.
+  But the last community version they served is now two major versions behind the commercial edition.
+
+Commercial support for Kafka is available from Confluent, should we need it.
+
+The Engineering Facilities Database
+-----------------------------------
+
+DDS requires running complex extra software (ts_salkafka, an in-house package) in order to copy all messages from DDS to Kafka, so we can ingest them into the Engineering Facilities Database (EFD).
+If we use Kafka then we can eliminate this extra software.
+
+Performance
+===========
+
+I measured performance by writing 10,000 messages for a specified topic in one process, and reading the same in another.
+Thus these tests measure latency, write speed, and whether the ts_salobj reader can keep up with the flood of messages, but they do not measure maximum read speed.
+
+I measured latency by comparing the time it received the message from the salobj topic against the private_sndStamp field.
+This measures latency when using salobj, rather than the theoretical best latency offered the underlying messaging system.
+
+One can get a good measure of DDS's intrinsic latency by comparing the private_rcvStamp field (which is set based on DDS's own report of when it read the message into its internal queue) against private_sndStamp.
+I performed that test and found maximum latency of a few ms (even when salobj was losing messages).
```

Comment:
I removed this. I see no point in reporting latency when running at maximum write speed, since it artificially slows down if the reader falls behind. That said, the latency was low for all cases except a few, so it's not a big deal.